src/transformers/generation/utils.py

```diff
@@ -3262,7 +3263,10 @@ def beam_search(
                 outputs, model_kwargs, is_encoder_decoder=self.config.is_encoder_decoder
             )
             if model_kwargs["past_key_values"] is not None:
+                using_new_cache = isinstance(model_kwargs["past_key_values"], Cache)
                 model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
+                if using_new_cache:
+                    model_kwargs["past_key_values"] = DynamicCache.from_legacy_cache(model_kwargs["past_key_values"])
```

Comment:
`DynamicCache` is an instance of `Cache`, self._reorder_cache` returns 
```suggestion
                using_new_cache = isinstance(model_kwargs["past_key_values"], DynamicCache)
                model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
                if using_new_cache:
                    model_kwargs["past_key_values"] = DynamicCache.from_legacy_cache(model_kwargs["past_key_values"])
```
If we were using a `SinkCache` then the new cache's type will be lost + only the `DynamicCache` will have the `from_legacy_cache` no?

Reply:
Same comment as: https://github.com/huggingface/transformers/pull/26681/files#r1412782842